### PR TITLE
src/fabric.c: Add loading of DL providers by their names

### PR DESCRIPTION
**Pull request by JuliaRS**
Library is caught from LD_LIBRARY_PATH if prov libs built as fi<provname>-fi.so

1) Why is UDP provider spelled in capital letters?
2) Why is psm2 provider built in fipsmx2-fi.so dl library? The symbol "x" in "psmx2" does not comply with the general rules for generating a dl library name.

### CODE REVIEW

File: src/fabric.c [hunk 1]
```diff
@@ -563,32 +586,50 @@ static void ofi_ini_dir(const char *dir)
...
 }
+
+static void ofi_find_prov_libs(void)
+{
+	const char* lib_prefix = "lib";
+	const char* lib_postfix = "-fi.so";
+	struct ofi_prov *prov;
+
+	/* Try to find the libraries that are visible by their name
+		* (LD_LIBRARY_PATH, system paths) */
```
Comment (shefty) [inline src/fabric.c:608]: Please fix-up indentation and move comment above the function.  I would change up the comment slightly, e.g.:
```
/* Search standard system library paths (i.e. LD_LIBRARY_PATH) for known DL provider
 * libraries.
 */
```

File: src/fabric.c [hunk 2]
```diff
@@ -563,32 +586,50 @@ static void ofi_ini_dir(const char *dir)
...
+	for (prov = prov_head; prov; prov = prov->next) {
+		char* lib;
+
+		if (!prov->prov_name)
+		continue;
```
Comment (shefty) [inline src/fabric.c:613]: indentation is off

File: src/fabric.c [hunk 3]
```diff
@@ -563,32 +586,50 @@ static void ofi_ini_dir(const char *dir)
...
+static void ofi_find_prov_libs(void)
+{
+	const char* lib_prefix = "lib";
+	const char* lib_postfix = "-fi.so";
+	struct ofi_prov *prov;
+
+	/* Try to find the libraries that are visible by their name
+		* (LD_LIBRARY_PATH, system paths) */
+	for (prov = prov_head; prov; prov = prov->next) {
+		char* lib;
```
Comment (shefty) [inline src/fabric.c:610]: Please declare all variables at the top of the function.  It makes them easier to locate.

File: src/fabric.c [hunk 4]
```diff
@@ -563,32 +586,50 @@ static void ofi_ini_dir(const char *dir)
...
+
+		if (!prov->prov_name)
+		continue;
+		
+		char* short_prov_name;
```
Comment (shefty) [inline src/fabric.c:615]: Move to top

File: src/fabric.c [hunk 5]
```diff
@@ -563,32 +586,50 @@ static void ofi_ini_dir(const char *dir)
...
 libdl_done:
 	while (n-- > 0)
 		free(liblist[n]);
 	free(liblist);
 }
+
+static void ofi_find_prov_libs(void)
+{
+	const char* lib_prefix = "lib";
+	const char* lib_postfix = "-fi.so";
```
Comment (shefty) [inline src/fabric.c:604]: Re-use FI_LIB_SUFFIX, which ends up being "fi.so".  The "-" will need to be added.

File: src/fabric.c [hunk 6]
```diff
@@ -563,32 +586,50 @@ static void ofi_ini_dir(const char *dir)
...
+		if (ofi_has_util_prefix(prov->prov_name)) {
+			short_prov_name = prov->prov_name + strlen(OFI_UTIL_PREFIX);		
+		} else {
+			short_prov_name = prov->prov_name;
+		}
+
+		if (asprintf(&lib, "%s%s%s", lib_prefix,
+			short_prov_name, lib_postfix) < 0) {
+			FI_WARN(&core_prov, FI_LOG_CORE,
+				"asprintf failed to allocate memory\n");
```
Comment (shefty) [inline src/fabric.c:625]: Need to continue here to avoid trying to access invalid memory lib

**Comment by JuliaRS** [comment]
@shefty , why is psm2 provider built in fipsmx2-fi.so dl library? The symbol "x" in "psmx2" does not comply with the general rules for generating a dl library name.
We should define the rules for the formation of library names and follow them.

**Comment by JuliaRS** [comment]
> Thanks - this is close, just needs some slight code cleanup.

@shefty , slight code cleanup is done.

**Pull request merged**